### ISSUE TITLE
Rendering Issue with Qt 5.9

### ISSUE THREAD

**Issue by oracle3001** [leafcode/libsleek#3]
Libsleek worked fine upto Qt 5.8, however with 5.9 it doesn't show the minimize / maximize / close buttons produces a strange extra area area which when you change the move the window get strange flickering. Also you can't grab the each of the window and resize it.

![libsleek_render_problem](https://user-images.githubusercontent.com/20092212/26839489-15cc0492-4adb-11e7-9a10-256be093ddd6.png)

**Comment by martinmolin** [comment]
Ah bummer, I haven't used this myself in a long time. I wonder what broke with Qt 5.9.. I'll install it and check it out myself!

**Comment by martinmolin** [comment]
Getting the same issue as you with 5.9, I'll investigate.

**Comment by martinmolin** [comment]
"winrect" here seems to be off in 5.9. Not sure what would case that. Setting border_width to 15 (current is 5) allows resizing to work in 5.9. But then the resize area will be too large if anything prior to 5.9 is used, and also the transparency issue is still there. I believe that finding the cause for this might solve the whole transparency/missing buttons issue since it seems like it is "outside" of its own area.

WINDOWPLACEMENT wp;
wp.length = sizeof( WINDOWPLACEMENT );
GetWindowPlacement( getParentWindow(), &wp );
if ( wp.showCmd == SW_MAXIMIZE )
    return false;

RECT winrect;
GetWindowRect((HWND)window()->winId(), &winrect);

**Comment by dfct** [comment]
I can save you further debugging, just had to fix & figure this out myself - https://github.com/dfct/TrueFramelessWindow/issues/4

Qt 5.9 introduced a patch that changes how they calculate window frame sizes. If you want custom frame margins (say, zero margins), you need to have FramelessWindowHint set *before* the window is created such as via winId(). There's a whole bunch of new code and probably alternative methods, but this was the best I could figure out.

**Comment by martinmolin** [comment]
Thank you, it solved the rendering but broke my resize code. Not a
silverbullet but definitely in the right direction!

2017-07-06 17:51 GMT+02:00 Ian Bannerman <notifications@github.com>:

**Comment by francislavoie** [comment]
Just a bump here, I was hoping to use this in one of my current projects which uses Qt 5.9. Haven't tried it yet, but is there a chance of getting a fix in?

**Comment by martinmolin** [comment]
In the current state I'd encourage you to see if you can find an alternative, or try to investigate the issue. I haven't had time to look at it more than the initial investigation. I might be able to take a look again this coming weekend but I cannot guarantee that I can solve it!

**Comment by martinmolin** [comment]
You could try https://github.com/dfct/TrueFramelessWindow I think he solved the issue!

**Comment by xoxoj** [comment]
@martinmolin  why ,just can open modal window?

**Comment by francislavoie** [comment]
@martinmolin I tried TrueFramelessWindow as you suggested, but it looks like this by default ![](https://i.imgur.com/fVx40tM.png)

Which definitely isn't what I'm looking for. I'll this lib a shot, I'll see if I can fix the resize issues while I'm at it, if it works for me.

**Comment by martinmolin** [comment]
@xoxoj You can open regular windows as well, what you see there is just a small sample application that demonstrate how to use the library :)

@francislavoie Dang it, I will try to look at this issue tomorrow if I find some time!

**Comment by francislavoie** [comment]
Nah, don't worry for my sake, I ended up finding another solution.

The reason I wanted a wrapper for frameless on windows is that I was seeing a bug completely mucking up the rendering of QML on windows, unless frameless is turned on.

I found that I can actually just wrap the QML window with a window container, and that actually solved the rendering problems for me.

```C++
    // Start main window
    QQmlApplicationEngine engine(&app);
    engine.load(QUrl("qrc:///qml/MainWindow.qml"));
    QObject *top = engine.rootObjects().value(0);

    // Set up the QQuickWindow with the QML
    QQuickWindow *qquickwin = qobject_cast<QQuickWindow *>(top);

    // Wrap the QQuickWindow in a QWidget window container
    QWidget *win = QWidget::createWindowContainer(qquickwin);
    win->resize(800, 480);

    if (!win) {
        qWarning("Error: root is bad");
        return -1;
    }
    win->show();
```

**Comment by martinmolin** [comment]
Oh, I see! Awesome that you managed to solve the issue, good luck with your project!
